### PR TITLE
[complex32] fft support (cuda only)

### PR DESCRIPTION
**Pull request by kshitij12345**
`half` and `complex32` support for `torch.fft.{fft, fft2, fftn, hfft, hfft2, hfftn, ifft, ifft2, ifftn, ihfft, ihfft2, ihfftn, irfft, irfft2, irfftn, rfft, rfft2, rfftn}`

* We only add support for `CUDA` as `cuFFT` supports these precision.
* We still error out on `CPU` and `ROCm` as their respective backends don't support this precision

For `cuFFT` following are the constraints for these precisions
* Minimum GPU architecture is SM_53
* Sizes are restricted to powers of two only
* Strides on the real part of real-to-complex and complex-to-real transforms are not supported
* More than one GPU is not supported
* Transforms spanning more than 4 billion elements are not supported

Ref: https://docs.nvidia.com/cuda/cufft/#half-precision-transforms

TODO:
* [x] Update docs about the restrictions
* [x] Check the correct way to check for `hip` device. (seems like `device.is_cuda()` is true for hip as well) (Thanks @peterbell10 )

Ref  for second point in TODO:https://github.com/pytorch/pytorch/blob/e424e7d214c02a79aab84d2d096b5b014f73ccd7/aten/src/ATen/native/SpectralOps.cpp#L31

### CODE REVIEW

File: test/test_spectral_ops.py [hunk 1]
```diff
@@ -430,7 +492,11 @@ def test_fftn_round_trip(self, device, dtype):
                               (torch.fft.ihfftn, torch.fft.hfftn)]
 
         for input_ndim, dim in transform_desc:
-            shape = itertools.islice(itertools.cycle(range(4, 9)), input_ndim)
+            if dtype in (torch.half, torch.complex32):
+                # cuFFT supports powers of 2 for half and complex half precision
```
Comment (peterbell10) [inline test/test_spectral_ops.py:496]: Please add a test that power of 2 values raise an appropriate error message.

File: aten/src/ATen/native/cuda/SpectralOps.cu [hunk 2]
```diff
@@ -75,10 +76,11 @@ C10_LAUNCH_BOUNDS_1(cuda::detail::CUDA_NUM_THREADS)
 __global__ void _fft_conjugate_copy_kernel(
     int64_t numel, scalar_t * out_data, const scalar_t * in_data,
     inp_calc_t ic, out_calc_t oc) {
+  using comp_t = at::opmath_type<scalar_t>;
   CUDA_KERNEL_LOOP_TYPE(index, numel, int64_t) {
     auto in_offset = ic.get(index)[0];
     auto out_offset = oc.get(index)[0];
-    out_data[out_offset] = std::conj(in_data[in_offset]);
+    out_data[out_offset] = std::conj(static_cast<comp_t>(in_data[in_offset]));
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/SpectralOps.cu]: It should be. Thanks :)

**Comment by anjali411** [comment]
> Strides on the real part of real-to-complex and complex-to-real transforms are not supported
can you elaborate what this means and how might this surface in user code?

File: test/test_spectral_ops.py [hunk 3]
```diff
@@ -284,13 +319,17 @@ def test_fft_round_trip(self, device, dtype):
                 }
 
                 y = backward(forward(x, **kwargs), **kwargs)
+                if x.dtype is torch.half and y.dtype is torch.complex32:
+                    # Since type promotion currently doesn't work with complex32
```
Comment (anjali411) [inline test/test_spectral_ops.py:323]: we should just add a small PR to add type promotion support for complex32. it should be fairly straightforward -- you just need to update the type promotion table

File: test/test_spectral_ops.py [hunk 4]
```diff
@@ -342,17 +386,28 @@ def test_fft_type_promotion(self, device, dtype):
...
         }
-        R = torch.fft.hfft(t)
+        if dtype in (torch.half, torch.complex32):
+            # cuFFT supports powers of 2 for half and complex half precision
+            # NOTE: With hfft and default args where output_size n=2*(input_dim - 1),
+            # we make sure that logical fft size is a power of two.
+            x = torch.randn(65, device=device, dtype=dtype)
+            R = torch.fft.hfft(x)
+        else:
+            # cuFFT supports powers of 2 for half and complex half precision
```
Comment (anjali411) [inline test/test_spectral_ops.py:403]: but we don't enter the else for complex32

File: torch/fft/__init__.py [hunk 5]
```diff
@@ -20,14 +20,17 @@
...
     The Fourier domain representation of any real signal satisfies the
     Hermitian property: `X[i] = conj(X[-i])`. This function always returns both
     the positive and negative frequency terms even though, for real inputs, the
     negative frequencies are redundant. :func:`~torch.fft.rfft` returns the
     more compact one-sided representation where only the positive frequencies
     are returned.
 
+Note:
+    Supports torch.half and torch.chalf on CUDA with GPU Architecture SM53 or greater.
+    However it only supports powers of 2 signal length along the dimenion of FFT.
```
Comment (anjali411) [inline torch/fft/__init__.py:32]:
```suggestion
    However it only supports powers of 2 signal length along the dimensions of FFT.
```

File: test/test_spectral_ops.py (hunk 4 above)
Comment (kshitij12345) [reply test/test_spectral_ops.py:403]: Oops. Will remove that.

File: test/test_spectral_ops.py [hunk 6]
```diff
@@ -342,17 +386,28 @@ def test_fft_type_promotion(self, device, dtype):
...
             torch.float: torch.float,
             torch.double: torch.double,
+            torch.complex32: torch.half,
             torch.complex64: torch.float,
             torch.complex128: torch.double,
         }
-        R = torch.fft.hfft(t)
+        if dtype in (torch.half, torch.complex32):
+            # cuFFT supports powers of 2 for half and complex half precision
+            # NOTE: With hfft and default args where output_size n=2*(input_dim - 1),
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply test/test_spectral_ops.py:398]: Sorry it should be `input_size`.

Ref `n` argument from the docs: https://pytorch.org/docs/stable/generated/torch.fft.hfft.html#torch.fft.hfft

File: torch/fft/__init__.py (hunk 5 above)
Comment (kshitij12345) [reply torch/fft/__init__.py:32]: Thanks!

**Comment by kshitij12345** [comment]
> does this just mean that non-contiguous tensors are not supported?

This is taken care of in

https://github.com/pytorch/pytorch/blob/b557e102d80393fdf1abaaef3454f8e412bc7b2f/aten/src/ATen/native/cuda/CuFFTPlanCache.h#L268-L285

It takes care of cloning the data, if input strides are not 1.

Note that the check is for `dtype == kHalf`, as only real scalar value is expected at that point of code. (Below code corresponds to the point where corresponding real value is taken)

https://github.com/pytorch/pytorch/blob/b557e102d80393fdf1abaaef3454f8e412bc7b2f/aten/src/ATen/native/cuda/SpectralOps.cpp#L260-L262

File: test/test_spectral_ops.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kshitij12345) [reply test/test_spectral_ops.py]: Updated.

File: torch/fft/__init__.py [hunk 7]
```diff
@@ -203,14 +218,17 @@
...
     The Fourier domain representation of any real signal satisfies the
     Hermitian property: ``X[i_1, ..., i_n] = conj(X[-i_1, ..., -i_n])``. This
     function always returns all positive and negative frequency terms even
     though, for real inputs, half of these values are redundant.
     :func:`~torch.fft.rfftn` returns the more compact one-sided representation
     where only the positive frequencies of the last dimension are returned.
 
+Note:
+    Supports torch.half and torch.chalf on CUDA with GPU Architecture SM53 or greater.
+    However it only supports powers of 2 signal size along the dimensions of FFT.
```
Comment (peterbell10) [inline torch/fft/__init__.py:230]: "dimensions of the FFT" is a bit unclear, can we say "in every transformed dimension". For hermitian signals, also be sure to say that `2(n-1)` must be a power to two in the symmetric dimension.

File: test/test_spectral_ops.py [hunk 8]
```diff
@@ -407,8 +484,14 @@ def test_reference_nd(self, device, dtype, op):
 
     @skipCPUIfNoFFT
     @onlyNativeDeviceTypes
-    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
+    @toleranceOverride({
+        torch.half : tol(1e-2, 1e-2),
+        torch.chalf : tol(1e-2, 1e-2),
+    })
+    @dtypes(torch.float, torch.double, torch.complex32, torch.complex64, torch.complex128)
```
Comment (peterbell10) [inline test/test_spectral_ops.py:491]: Was `torch.half` left out intentionally?

**Review by peterbell10 — COMMENTED** [review]
>Check the correct way to check for hip device. (seems like device.is_cuda() is true for hip as well)

I think `at::detail::getCUDAHooks().compiledWithMIOpen()`  will allow you to access `AT_ROCM_ENABLED()` from cpu code. You might want to add a separate hook though, on the off chance a build config with ROCm, but not MIOpen gets added.

File: test/test_spectral_ops.py (hunk 8 above)
Comment (kshitij12345) [reply test/test_spectral_ops.py:491]: Great catch. Thanks!

File: test/test_spectral_ops.py (hunk 3 above)
Comment (kshitij12345) [reply test/test_spectral_ops.py:228]: Noted.

File: aten/src/ATen/native/SpectralOps.cpp [hunk 9]
```diff
@@ -28,14 +28,19 @@ ScalarType promote_type_fft(ScalarType type, bool require_complex) {
     type = c10::typeMetaToScalarType(c10::get_default_dtype());
   }
 
-  TORCH_CHECK(type == kFloat || type == kDouble, "Unsupported dtype ", type);
+  if (device.is_cuda() && !at::detail::getCUDAHooks().hasROCM()) {
+    TORCH_CHECK(type == kHalf || type == kFloat || type == kDouble, "Unsupported dtype ", type);
```
Comment (anjali411) [inline aten/src/ATen/native/SpectralOps.cpp:32]: do we usually specify the rocm support in the docs?

File: test/test_spectral_ops.py [hunk 10]
```diff
@@ -110,6 +112,20 @@ def _stft_reference(x, hop_length, window):
         X[:, m] = torch.fft.fft(slc * window)
     return X
 
+
+def skip_helper(device, dtype):
```
Comment (anjali411) [inline test/test_spectral_ops.py:116]: nit-> `skip_helper_for_fft`

Comment (anjali411) [inline test/test_spectral_ops.py:116]: nit-> `skip_helper_for_fft`

File: test/test_spectral_ops.py [hunk 11]
```diff
@@ -110,6 +112,20 @@ def _stft_reference(x, hop_length, window):
...
+    device_type = torch.device(device).type
+    if dtype not in (torch.half, torch.complex32):
+        return
+
+    if device_type == 'cpu':
+        raise unittest.SkipTest("half and complex32 not supported on CPU")
+    if TEST_WITH_ROCM:
+        raise unittest.SkipTest("half and complex32 not supported on ROCM")
```
Comment (anjali411) [inline test/test_spectral_ops.py:124]:
```suggestion
        raise unittest.SkipTest("half and complex32 are not supported on ROCM")
```

Comment (anjali411) [inline test/test_spectral_ops.py:124]:
```suggestion
        raise unittest.SkipTest("half and complex32 are not supported on ROCM")
```

File: test/test_spectral_ops.py [hunk 12]
```diff
@@ -110,6 +112,20 @@ def _stft_reference(x, hop_length, window):
...
     return X
 
+
+def skip_helper(device, dtype):
+    device_type = torch.device(device).type
+    if dtype not in (torch.half, torch.complex32):
+        return
+
+    if device_type == 'cpu':
+        raise unittest.SkipTest("half and complex32 not supported on CPU")
```
Comment (anjali411) [inline test/test_spectral_ops.py:122]:
```suggestion
        raise unittest.SkipTest("half and complex32 are not supported on CPU")
```

Comment (anjali411) [inline test/test_spectral_ops.py:122]:
```suggestion
        raise unittest.SkipTest("half and complex32 are not supported on CPU")
```

File: test/test_spectral_ops.py [hunk 13]
```diff
@@ -157,20 +173,39 @@ def test_reference_1d(self, device, dtype, op):
 
     @skipCPUIfNoFFT
     @onlyNativeDeviceTypes
-    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
+    @toleranceOverride({
+        torch.half : tol(1e-2, 1e-2),
+        torch.chalf : tol(1e-2, 1e-2),
```
Comment (anjali411) [inline test/test_spectral_ops.py:178]: is this standard for the rest of our tests for `half` and `chalf`?

Comment (anjali411) [inline test/test_spectral_ops.py:178]: is this standard for the rest of our tests for `half` and `chalf`?

File: torch/fft/__init__.py [hunk 14]
```diff
@@ -20,14 +20,17 @@
 Computes the one dimensional discrete Fourier transform of :attr:`input`.
 
 Note:
-
```
Comment (anjali411) [inline torch/fft/__init__.py:23]: spurious line deletion

Comment (anjali411) [inline torch/fft/__init__.py:23]: spurious line deletion

File: test/test_spectral_ops.py [hunk 15]
```diff
@@ -269,9 +323,32 @@ def test_fft_type_promotion(self, device, dtype):
...
+        elif dtype is torch.half and device_type == 'cuda' and not SM53OrLater:
+            err_msg = "cuFFT doesn't support signals of half type with compute capability less than SM_53"
+        else:
+            err_msg = "Unsupported dtype "
+        with self.assertRaisesRegex(RuntimeError, err_msg):
+            op(sample.input, *sample.args, **sample.kwargs)
+
+    @onlyNativeDeviceTypes
+    @ops(spectral_funcs, allowed_dtypes=(torch.half, torch.chalf))
+    def test_fft_half_and_chalf_not_power_of_two_error(self, device, dtype, op):
```
Comment (anjali411) [inline test/test_spectral_ops.py:339]: nice!

**Review by anjali411 — APPROVED** [review]
LGTM! I'll let @peterbell10 take a final look at it.

**Review by peterbell10 — APPROVED** [review]
@kshitij12345 can you rebase, some of the CI is failing on unrelated tests.

File: torch/fft/__init__.py (hunk 14 above)
Comment (kshitij12345) [reply torch/fft/__init__.py:23]: I think these new lines were spurious.

https://github.com/pytorch/pytorch/blob/3c2e0dc6574e1ddc1dda16067200d5b491a18fa7/torch/fft/__init__.py#L357-L368

File: test/test_spectral_ops.py (hunk 13 above)
Comment (kshitij12345) [reply test/test_spectral_ops.py:178]: These for consistent for fft. For other test like `test_complex_half_reference_testing` we use default value for `self.assertEqual` which defaults to `atol=1e-3, rtol=1e-5`

**Comment by kshitij12345** [comment]
Failed CI looks irrelevant.

**Pull request closed without merging**